sphinx_gallery/gen_gallery.py

```diff
@@ -329,6 +343,36 @@ def sumarize_failing_examples(app, exception):
                          "\n" + "-" * 79)
 
 
+def collect_gallery_files(examples_dirs):
+    """Collect python files from the gallery example directories."""
+    files = []
+    for example_dir in examples_dirs:
+        for root, dirnames, filenames in os.walk(example_dir):
+            for filename in filenames:
+                if filename.endswith('.py'):
+                    files.append(os.path.join(root, filename))
+    return files
+
+
+def check_duplicate_filenames(files):
+    """Check for duplicate filenames across gallery directories."""
```

Comment:
Why is this a problem? Don't examples get a full path name as their cross-reference name, e.g., `sphx_glr_examples_directory_plot_demo` for `examples/directory/plot_demo.py`?

This warning will break Matplotlib doc builds since we turn warnings into errors and there are currently some duplicate example files.

Reply:
Check out the discussion above starting [here](https://github.com/sphinx-gallery/sphinx-gallery/pull/244#issuecomment-335534068). Basically examples can break each other unexpectedly because of how files are named by `Sphinx` (not `sphinx-gallery`), so a warning seems justified to me.

@choldgraf at the time also said:

> I just checked matplotlib and of their 537 examples, 5 had overlapping names. For the tutorials there were no overlapping names. 

To me the right thing to do is thus to fix the 5 overlapping names at the `matplotlib` end.

Reply:
yeah I am +1 on renaming files to have non-overlapping names...happy to make that PR.